hudi-utilities/src/main/java/org/apache/hudi/utilities/multitable/MultiTableServiceUtils.java

```diff
@@ -0,0 +1,167 @@
+import org.apache.hudi.common.config.SerializableConfiguration;
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.exception.HoodieException;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.spark.api.java.JavaSparkContext;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.stream.Collectors;
+
+import static org.apache.hudi.common.table.HoodieTableMetaClient.METAFOLDER_NAME;
+
+/**
+ * Utils for executing multi-table services
+ */
+public class MultiTableServiceUtils {
+
+  public static class Constants {
+    public static final String TABLES_TO_BE_SERVED_PROP = "hoodie.tableservice.tablesToServe";
+
+    public static final String COMMA_SEPARATOR = ",";
+
+    private static final int DEFAULT_LISTING_PARALLELISM = 1500;
+  }
+
+  public static List<String> getTablesToBeServedFromProps(TypedProperties properties) {
+    String combinedTablesString = properties.getString(Constants.TABLES_TO_BE_SERVED_PROP);
+    if (combinedTablesString == null) {
+      return new ArrayList<>();
+    }
+    String[] tablesArray = combinedTablesString.split(Constants.COMMA_SEPARATOR);
+    return Arrays.asList(tablesArray);
+  }
+
+  public static List<String> findHoodieTablesUnderPath(JavaSparkContext jsc, String pathStr) {
+    Path rootPath = new Path(pathStr);
+    SerializableConfiguration conf = new SerializableConfiguration(jsc.hadoopConfiguration());
+    if (isHoodieTable(rootPath, conf.get())) {
+      return Collections.singletonList(pathStr);
+    }
+
+    HoodieSparkEngineContext engineContext = new HoodieSparkEngineContext(jsc);
+    List<String> hoodieTablePaths = new CopyOnWriteArrayList<>();
+    List<Path> pathsToList = new CopyOnWriteArrayList<>();
+    pathsToList.add(rootPath);
+    int listingParallelism = Math.min(Constants.DEFAULT_LISTING_PARALLELISM, pathsToList.size());
+
+    while (!pathsToList.isEmpty()) {
+      // List all directories in parallel
+      List<FileStatus[]> dirToFileListing = engineContext.map(pathsToList, path -> {
+        FileSystem fileSystem = path.getFileSystem(conf.get());
+        return fileSystem.listStatus(path);
+      }, listingParallelism);
+      pathsToList.clear();
+
+      // if current dictionary contains meta folder(.hoodie), add it to result. Otherwise, add it to queue
+      List<FileStatus> dirs = dirToFileListing.stream().flatMap(Arrays::stream)
+          .filter(FileStatus::isDirectory)
+          .collect(Collectors.toList());
+
+      if (!dirs.isEmpty()) {
+        List<Pair<FileStatus, Integer>> dirResults = engineContext.map(dirs, fileStatus -> {
+          if (isHoodieTable(fileStatus.getPath(), conf.get())) {
```

Comment:
what does 0, 1, 2 means, looks like magic number?

Reply:
Nice catch~ It's not a good design that uses hard-coded magic number, already updated the magic number to meaningful enum constants. 

Updated to: 
```java
  /**
   * Type of directories when searching hoodie tables under path
   */
  enum DirType {
    HOODIE_TABLE, // previous 0
    NORMAL_DIR, // previous 1 
    META_FOLDER  // previous 2 
  }
```